File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-4.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} subRoot
     * @return {boolean}
     */
    isSubtree(root, subRoot) {
        
        function dfs(node){
            if(!node) return false;

            if(sameTree(node, subRoot)) return true;

            return dfs(node.left) || dfs(node.right);
        }

        function sameTree(node, sub){
            if(!node && !sub) return true;

            if(!node || !sub) return false;

            if(node.val !== sub.val) return false;

            return sameTree(node.left, sub.left) && sameTree(node.right, sub.right);
        }

        return dfs(root);
    }
}

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:

        if not root:
            return []
        q = deque()

        q.append(root)

        res = []



        while q:

            curr_len = len(q)

            lst = []

            for i in range(curr_len):
                curr = q.popleft()

                if curr.left:
                    q.append(curr.left) 
                if curr.right:
                    q.append(curr.right) 

                lst.append(curr.val)
            
            res.append(lst)


        return res




